src/mesh/stl_io.C

```diff
@@ -0,0 +1,358 @@
+#include <iomanip>
+#include <fstream>
+#include <vector>
+#include <cctype> // isspace
+
+#ifdef LIBMESH_HAVE_CXX11_REGEX
+#include <regex>
+#endif
+
+namespace libMesh
+{
+
+STLIO::STLIO (const MeshBase & mesh) :
+  MeshOutput<MeshBase>    (mesh),
+  _subdivide_second_order (true)
+{
+}
+
+
+
+STLIO::STLIO (MeshBase & mesh) :
+  MeshInput<MeshBase> (mesh),
+  MeshOutput<MeshBase>(mesh),
+  _subdivide_second_order (true)
+{
+}
+
+
+
+void STLIO::write (const std::string & fname)
+{
+  LOG_SCOPE("write()", "STLIO");
+
+  // Get a reference to the mesh
+  const MeshBase * mesh_ptr = &MeshOutput<MeshBase>::mesh();
+
+  libmesh_parallel_only(mesh_ptr->comm());
+
+  // If necessary, serialize to proc 0, which will do all the writing
+  std::unique_ptr<DistributedMesh> mesh_copy;
+  std::unique_ptr<MeshSerializer> serializer;
+  if (!mesh_ptr->is_serial())
+    {
+      mesh_copy = std::make_unique<DistributedMesh>(*mesh_ptr);
+      mesh_ptr = mesh_copy.get();
+      serializer = std::make_unique<MeshSerializer>(*mesh_copy, true, true);
+    }
+
+  const MeshBase & mesh = *mesh_ptr;
+
+  if (mesh.processor_id() != 0)
+    return;
+
+  // Open the output file stream
+  std::ofstream out_stream (fname.c_str());
+
+  out_stream << std::setprecision(this->ascii_precision());
+
+  // Make sure it opened correctly
+  if (!out_stream.good())
+    libmesh_file_error(fname.c_str());
+
+  // Write the header
+  out_stream << "solid " << this->_name << '\n';
+
+  // Write all our triangles
+  // scream if we see a non-triangle
+  for (auto elem : mesh.element_ptr_range())
+    {
+      // Get to this later
+      if (elem->type() == TRI6 || elem->type() == TRI7)
+        {
+          if (this->_subdivide_second_order)
+            libmesh_not_implemented();
+
+          libmesh_error_msg("Tried to write a non-linear triangle to an STL file");
+        }
+
+      libmesh_error_msg_if(elem->type() != TRI3,
+                           "Tried to write a non-triangle to an STL file");
+
+      auto n = (elem->point(1)-elem->point(0)).cross(elem->point(2)-elem->point(0));
+      n = n.unit();
+
+      out_stream << "facet normal " <<
+        n(0) << ' ' <<
+        n(1) << ' ' <<
+        n(2) << "\n"
+        "    outer loop\n"
+        "        vertex " <<
+        elem->point(0)(0) << ' ' <<
+        elem->point(0)(1) << ' ' <<
+        elem->point(0)(2) << "\n"
+        "        vertex " <<
+        elem->point(1)(0) << ' ' <<
+        elem->point(1)(1) << ' ' <<
+        elem->point(1)(2) << "\n"
+        "        vertex " <<
+        elem->point(2)(0) << ' ' <<
+        elem->point(2)(1) << ' ' <<
+        elem->point(2)(2) << "\n"
+        "    endloop\n"
+        "endfacet\n";
+    }
+
+  out_stream << "endsolid" << std::endl;
+}
+
+
+void STLIO::read (const std::string & filename)
+{
+  LOG_SCOPE("read()", "STLIO");
+
+#ifndef LIBMESH_HAVE_CXX11_REGEX
+  libmesh_not_implemented();  // What is your compiler?!?  Email us!
+  libmesh_ignore(filename);
+#else
```

Comment:
Should we make C++11 regex support officially required now? Looking at configure.ac, I see:
```
# --------------------------------------------------------------
# Test for optional modern C++ features, which libMesh offers shims
# for and/or which libMesh applications may want to selectively use.
# --------------------------------------------------------------
LIBMESH_TEST_CXX14_MAKE_UNIQUE
LIBMESH_TEST_CXX11_REGEX
```
and we obviously use `std::make_unique` freely everywhere now, so I think these could probably both be moved to the `ACSM_TEST_CXX_ALL` test, which is what sets the `have_cxx_all` flag. A topic for another PR though, either way.

Reply:
Yeah, that sounds good to me.

[cppreference claims](https://en.cppreference.com/w/cpp/compiler_support/11) that Embarcadero C++ Builder and Sun/Oracle C++ are the only libraries that didn't add regex support ... but [later claims](